src/main/kotlin/org/jitsi/nlj/rtp/TransportCcEngine.kt

```diff
@@ -0,0 +1,152 @@
+     * (that of the sender). This is technically not necessary and it's done for
+     * convenience.
+     */
+    private var localReferenceTime: Instant = NEVER
+
+    /**
+     * Holds a key value pair of the packet sequence number and an object made
+     * up of the packet send time and the packet size.
+     */
+    private val sentPacketDetails = LRUCache<Int, PacketDetail>(MAX_OUTGOING_PACKETS_HISTORY)
+
+    /**
+     * Called when an RTP sender has a new round-trip time estimate.
+     */
+    fun onRttUpdate(avgRttMs: Long, maxRttMs: Long) {
+        val now = clock.instant()
+        bandwidthEstimator.onRttUpdate(now, Duration.ofMillis(avgRttMs))
+    }
+
+    override fun rtcpPacketReceived(rtcpPacket: RtcpPacket, receivedTime: Long) {
+        if (rtcpPacket is RtcpFbTccPacket) {
+            tccReceived(rtcpPacket)
+        }
+    }
+
+    private fun tccReceived(tccPacket: RtcpFbTccPacket) {
+        val now = clock.instant()
+        var currArrivalTimestamp = Instant.ofEpochMilli(tccPacket.GetBaseTimeUs() / 1000)
+        if (remoteReferenceTime == NEVER) {
+            remoteReferenceTime = currArrivalTimestamp
+            localReferenceTime = now
+        }
+
+        for (receivedPacket in tccPacket) {
+            val packetDetail: PacketDetail?
+            val tccSeqNum = receivedPacket.seqNum
+            synchronized(sentPacketsSyncRoot) {
+                packetDetail = sentPacketDetails.remove(tccSeqNum)
+            }
+
+            if (packetDetail == null) {
+                if (receivedPacket.received) {
+                    logger.warn("Couldn't find packet detail for $tccSeqNum.")
+                }
+                continue
+            }
+
+            if (!receivedPacket.received) {
+                bandwidthEstimator.processPacketLoss(now, packetDetail.packetSendTime, tccSeqNum)
+                continue
+            }
+            val delta = Duration.of(receivedPacket.deltaTicks * 250L, ChronoUnit.MICROS)
+            currArrivalTimestamp += delta
+
+            val arrivalTimeInLocalClock = currArrivalTimestamp - Duration.between(localReferenceTime, remoteReferenceTime)
+
+            bandwidthEstimator.processPacketArrival(
+                now, packetDetail.packetSendTime, arrivalTimeInLocalClock, tccSeqNum, packetDetail.packetLength)
+        }
+    }
+
+    fun mediaPacketSent(tccSeqNum: Int, length: DataSize) {
+        synchronized(sentPacketsSyncRoot) {
+            val now = clock.instant()
+            sentPacketDetails.put(
+                tccSeqNum and 0xFFFF,
+                PacketDetail(length, now))
+        }
+    }
+
+    /**
+     * [PacketDetail] is an object that holds the
+     * length(size) of the packet in [packetLength]
+     * and the time stamps of the outgoing packet
+     * in [packetSendTime]
+     */
+    private data class PacketDetail internal constructor(internal var packetLength: DataSize, internal var packetSendTime: Instant)
```

Comment:
looks like `packetLength` and `packetSendTime` should be `val`s

Reply:
Yeah, that seems to have been an artifact of the automated Java-to-Kotlin conversion.